core/arch/arm/kernel/ftrace.c

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019, Linaro Limited
+ */
+
+#include <assert.h>
+#include <kernel/ftrace.h>
+#include <kernel/panic.h>
+#include <kernel/thread.h>
+#include <kernel/tee_ta_manager.h>
+#include <mm/mobj.h>
+#include <optee_rpc_cmd.h>
+#include <printk.h>
+#include <string.h>
+#include <trace.h>
+#include "elf_load_dyn.h"
+
+#define MIN_FTRACE_BUF_SIZE	1024
+#define MAX_HEADER_STRLEN	128
+
+void ta_fbuf_init(vaddr_t load_addr, struct tee_ta_session *s,
+		  struct elf_load_state *state)
+{
+	struct __ftrace_info *finfo = NULL;
+	struct ftrace_buf *fbuf = NULL;
+	size_t fbuf_size = 0;
+	uintptr_t val = 0;
+	int count = 0;
+	TEE_Result res = TEE_SUCCESS;
+
+	if (state->is_32bit) {
+		DMSG("Current ftrace doesn't support 32 bit TAs");
+		return;
+	}
+
+	res = elf_resolve_symbol(state, "__ftrace_info", &val);
+	if (res != TEE_SUCCESS || !val) {
+		DMSG("ftrace info not found");
+		return;
+	}
+
+	finfo = (struct __ftrace_info *)(val + load_addr);
```

Comment:
We need to check that this is a valid user space address buffer.

Reply:
Would check using `tee_mmu_is_vbuf_inside_ta_private()` suffice?

Reply:
Yes